runtime/array.c

```diff
@@ -303,24 +306,81 @@ CAMLprim value caml_make_array(value init)
 
 /* Blitting */
 
+/* [wo_memmove] copies [nvals] values from [src] to [dst]. If there is a single
+   domain running, then we use [memmove]. Otherwise, we copy one word at a
+   time. Since the [memmove] implementation does not guarantee that the writes
+   are always word-sized, we explicitly perform word-sized writes to avoid
+   mixed-mode accesses.
+
+   [MM]: TODO enforcing the memory model also requires fences. See note in
+   memory.c */
+static void wo_memmove (value* const dst, const value* const src,
+                        mlsize_t nvals)
+{
+  mlsize_t i;
+
+  if (caml_domain_alone ()) {
+    memmove (dst, src, nvals * sizeof (value));
+  } else {
+    if (dst < src) {
+      /* copy ascending */
+      for (i = 0; i < nvals; i++) dst[i] = src[i];
+    } else {
+      /* copy descending */
+      for (i = nvals; i > 0; i--) dst[i-1] = src[i-1];
+    }
+  }
+}
```

Comment:
A good C compiler might try to use vector instructions to move several words at a time, resulting in mixed-size concurrent accesses...  However, if I read `caml_modify` correctly, you need to use atomic stores of the "release" kind, which should prevent the writes from being coalesced into vector writes.